src/AspNet.Security.OAuth.SuperOffice/SuperOfficeAuthenticationHandler.cs

```diff
@@ -0,0 +1,201 @@
+        private async Task<string> ProcessIdTokenAndGetContactIdentifierAsync(
+            [NotNull] OAuthTokenResponse tokens,
+            [NotNull] AuthenticationProperties properties,
+            [NotNull] ClaimsIdentity identity)
+        {
+            string contextIdentifier = string.Empty;
+
+            string idToken = tokens.Response.RootElement.GetString("id_token");
+
+            if (Options.SaveTokens)
+            {
+                // Save id_token as well.
+                SaveIdToken(properties, idToken);
+            }
+
+            IEnumerable<Claim>? claims = null;
+
+            if (Options.ValidateTokens)
+            {
+                var tempClaimsPrincipal = await ValidateAsync(
+                    idToken,
+                    new TokenValidationParameters()
+                    {
+                        ValidAudience = Options.ClientId,
+                        ValidIssuer = Options.ClaimsIssuer
+                    });
+
+                claims = tempClaimsPrincipal.Claims;
+            }
+            else
+            {
+                // This is to satisfy CS8602, the dereference of a null reference on Options.SecurityTokenHandler.
+                if (Options.SecurityTokenHandler == null)
+                {
+                    throw new NullReferenceException("The options SecurityTokenHandler is null.");
+                }
+
+                var jwtSecurityToken = Options.SecurityTokenHandler.ReadJwtToken(idToken);
+                claims = jwtSecurityToken.Claims;
+            }
+
+            foreach (var claim in claims)
+            {
+                if (claim.Type == SuperOfficeAuthenticationConstants.ClaimNames.ContextIdentifier)
+                {
+                    contextIdentifier = claim.Value;
+                }
+
+                if (Options.IncludeIdTokenAsClaims)
+                {
+                    // May be possible same claim names from UserInformationEndpoint and IdToken.
+                    if (!identity.HasClaim(c => c.Type == claim.Type))
+                    {
+                        identity.AddClaim(claim);
+                    }
+                }
+            }
+
+            return contextIdentifier;
+        }
+
+        /// <summary>
+        /// Store id_token in <paramref name="properties"/> token collection.
+        /// </summary>
+        /// <param name="properties">Authentication properties.</param>
+        /// <param name="idToken">The id_token JWT.</param>
+        private static void SaveIdToken(
+            [NotNull] AuthenticationProperties properties,
+            [NotNull] string idToken)
+        {
+            if (!string.IsNullOrWhiteSpace(idToken))
+            {
+                // Save existing tokens, which are removed in StoreTokens method.
+                var tokens = properties.GetTokens().ToList();
+                tokens.Add(new AuthenticationToken() { Name = "id_token", Value = idToken });
+                properties.StoreTokens(tokens);
```

Comment:
It took me a minute to fully understand what this was doing as "Save" was used next to a "Get" method. I think it could be made clearer:

```suggestion
                // Get the currently available tokens
                var  = properties.GetTokens().ToList();
                
                // Add the extra token
                tokens.Add(new AuthenticationToken() { Name = "id_token", Value = idToken });
                
                // Overwrite store with original tokens with the new additional token
                properties.StoreTokens(tokens);
```

Reply:
I will commit your suggestion, but as the original comment states StoreTokens removes any existing tokens and therefore does not overwrite anything - it replaces.

Reply:
To me, "overwrite" and "replace" are synonyms.